File: packages/core/src/connectors/github.ts
/**
 * @fileoverview GitHub connector for interacting with GitHub via the gh CLI.
 *
 * SECURITY: This connector NEVER logs stdout to prevent token exposure.
 */

import { type ExecOptions, type ExecResult, exec, isExecError } from '../platform.js';

// ═══════════════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════════════

/** Options for creating an issue */
export interface IssueCreateOptions {
	/** Issue title */
	title: string;
	/** Issue body/description */
	body: string;
	/** Repository in owner/repo format (uses current repo if not specified) */
	repo?: string;
	/** Labels to apply to the issue */
	labels?: string[];
	/** Project to add the issue to */
	project?: string;
}

/** Filters for listing issues */
export interface IssueFilters {
	/** Filter by state */
	state?: 'open' | 'closed' | 'all';
	/** Filter by label */
	label?: string;
	/** Filter by assignee */
	assignee?: string;
	/** Maximum number of issues to return */
	limit?: number;
}

/** Represents a GitHub issue */
export interface Issue {
	/** Issue number */
	number: number;
	/** Issue title */
	title: string;
	/** Issue state */
	state: 'open' | 'closed';
	/** URL to the issue */
	url: string;
}

/** Options for creating a pull request */
export interface PrCreateOptions {
	/** PR title */
	title: string;
	/** PR body/description */
	body: string;
	/** Source branch */
	head: string;
	/** Target branch (defaults to default branch) */
	base?: string;
	/** Create as draft PR */
	draft?: boolean;
}

/** Represents a GitHub pull request */
export interface PullRequest {
	/** PR number */
	number: number;
	/** PR title */
	title: string;
	/** PR state */
	state: 'open' | 'closed' | 'merged';
	/** URL to the PR */
	url: string;
}

/** Options for GitHubConnector */
export interface GitHubConnectorOptions {
	/** Enable debug logging of commands (never logs stdout for security) */
	debug?: boolean;
}

// ═══════════════════════════════════════════════════════════════════════════════
// GITHUB CONNECTOR
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Build a `--name=value` flag. Keeping the value in the same argument means a value that
 * starts with `-` (e.g. a title of `--web`) can never be read as another flag.
 */
function flag(name: string, value: string): string {
	return `--${name}=${value}`;
}

/**
 * Connector for interacting with GitHub via the gh CLI.
 *
 * SECURITY: This connector NEVER logs stdout to prevent accidental
 * exposure of tokens or sensitive data from gh CLI output.
 *
 * @example
 * ```typescript
 * const gh = new GitHubConnector({ debug: true });
 *
 * // Create an issue
 * const issueNum = await gh.issueCreate({
 *   title: 'Bug report',
 *   body: 'Description of the bug',
 *   labels: ['bug']
 * });
 *
 * // List issues
 * const issues = await gh.issueList('owner/repo', { state: 'open', limit: 10 });
 * ```
 */
export class GitHubConnector {
	private options?: GitHubConnectorOptions;

	constructor(options?: GitHubConnectorOptions) {
		this.options = options;
	}

	/**
	 * Create a new issue.
	 *
	 * @param opts - Issue creation options
	 * @returns Issue number on success, or throws on failure
	 */
	async issueCreate(opts: IssueCreateOptions): Promise<number> {
		const cmd: string[] = [
			'gh',
			'issue',
			'create',
			flag('title', opts.title),
			flag('body', opts.body),
		];

		if (opts.repo) {
			cmd.push(flag('repo', opts.repo));
		}

		if (opts.labels && opts.labels.length > 0) {
			for (const label of opts.labels) {
				cmd.push(flag('label', label));
			}
		}

		if (opts.project) {
			cmd.push(flag('project', opts.project));
		}

		const result = await this.execGh(cmd);

		if (isExecError(result)) {
			throw new Error(`Failed to create issue: ${result.stderr || result.errorCode}`);
		}

		// Parse issue number from URL in stdout (e.g., https://github.com/owner/repo/issues/123)
		const match = result.stdout.match(/\/issues\/(\d+)/);
		if (!match?.[1]) {
			throw new Error('Could not parse issue number from gh output');
		}

		return parseInt(match[1], 10);
	}

	/**
	 * List issues from a repository.
	 *
	 * @param repo - Repository in owner/repo format
	 * @param filters - Optional filters
	 * @returns Array of issues
	 */
	async issueList(repo: string, filters?: IssueFilters): Promise<Issue[]> {
		const cmd: string[] = [
			'gh',
			'issue',
			'list',
			flag('repo', repo),
			flag('json', 'number,title,state,url'),
		];

		if (filters?.state) {
			cmd.push(flag('state', filters.state));
		}

		if (filters?.label) {
			cmd.push(flag('label', filters.label));
		}

		if (filters?.assignee) {
			cmd.push(flag('assignee', filters.assignee));
		}

		if (filters?.limit) {
			cmd.push(flag('limit', String(filters.limit)));
		}

		const result = await this.execGh(cmd);

		if (isExecError(result)) {
			throw new Error(`Failed to list issues: ${result.stderr || result.errorCode}`);
		}

		const issues = JSON.parse(result.stdout) as Array<{
			number: number;
			title: string;
			state: string;
			url: string;
		}>;

		return issues.map((issue) => ({
			number: issue.number,
			title: issue.title,
			state: issue.state.toLowerCase() as 'open' | 'closed',
			url: issue.url,
		}));
	}

	/**
	 * Create a new pull request.
	 *
	 * @param opts - PR creation options
	 * @returns PR number on success, or throws on failure
	 */
	async prCreate(opts: PrCreateOptions): Promise<number> {
		const cmd: string[] = [
			'gh',
			'pr',
			'create',
			flag('title', opts.title),
			flag('body', opts.body),
			flag('head', opts.head),
		];

		if (opts.base) {
			cmd.push(flag('base', opts.base));
		}

		if (opts.draft) {
			cmd.push('--draft');
		}

		const result = await this.execGh(cmd);

		if (isExecError(result)) {
			throw new Error(`Failed to create PR: ${result.stderr || result.errorCode}`);
		}

		// Parse PR number from URL in stdout (e.g., https://github.com/owner/repo/pull/123)
		const match = result.stdout.match(/\/pull\/(\d+)/);
		if (!match?.[1]) {
			throw new Error('Could not parse PR number from gh output');
		}

		return parseInt(match[1], 10);
	}

	/**
	 * List pull requests from a repository.
	 *
	 * @param repo - Repository in owner/repo format
	 * @param filters - Optional filters (reuses IssueFilters for similar fields)
	 * @returns Array of pull requests
	 */
	async prList(repo: string, filters?: IssueFilters): Promise<PullRequest[]> {
		const cmd: string[] = [
			'gh',
			'pr',
			'list',
			flag('repo', repo),
			flag('json', 'number,title,state,url'),
		];

		if (filters?.state && filters.state !== 'all') {
			cmd.push(flag('state', filters.state));
		}

		if (filters?.label) {
			cmd.push(flag('label', filters.label));
		}

		if (filters?.assignee) {
			cmd.push(flag('assignee', filters.assignee));
		}

		if (filters?.limit) {
			cmd.push(flag('limit', String(filters.limit)));
		}

		const result = await this.execGh(cmd);

		if (isExecError(result)) {
			throw new Error(`Failed to list PRs: ${result.stderr || result.errorCode}`);
		}

		const prs = JSON.parse(result.stdout) as Array<{
			number: number;
			title: string;
			state: string;
			url: string;
		}>;

		return prs.map((pr) => ({
			number: pr.number,
			title: pr.title,
			state: pr.state.toLowerCase() as 'open' | 'closed' | 'merged',
			url: pr.url,
		}));
	}

	/**
	 * Execute a gh command.
	 *
	 * SECURITY: This method enables debug logging for commands but
	 * NEVER logs stdout to prevent token exposure.
	 */
	private execGh(cmd: string[]): Promise<ExecResult> {
		// Only pass debug flag for command logging, never log stdout
		const execOptions: ExecOptions = {};
		if (this.options?.debug) {
			execOptions.debug = true;
		}
		return exec(cmd, execOptions);
	}
}

File: python/src/afd/connectors/github.py
"""GitHub connector for interacting with GitHub via the gh CLI.

SECURITY: This connector NEVER logs stdout to prevent token exposure.

Example:
    >>> from afd.connectors.github import GitHubConnector
    >>> gh = GitHubConnector(GitHubConnectorOptions(debug=True))
    >>> issues = await gh.issue_list("owner/repo", IssueFilters(state="open"))
"""

import json
from typing import Literal, Optional

from pydantic import BaseModel, Field

from afd.platform import ExecOptions, exec_command, is_exec_error


# ═══════════════════════════════════════════════════════════════════════════════
# TYPES
# ═══════════════════════════════════════════════════════════════════════════════


class IssueCreateOptions(BaseModel):
    """Options for creating a GitHub issue.

    Attributes:
        title: Issue title.
        body: Issue body/description.
        repo: Repository in owner/repo format (uses current repo if not specified).
        labels: Labels to apply to the issue.
        project: Project to add the issue to.
    """

    title: str
    body: str
    repo: Optional[str] = None
    labels: Optional[list[str]] = None
    project: Optional[str] = None


class IssueFilters(BaseModel):
    """Filters for listing GitHub issues.

    Attributes:
        state: Filter by state (open, closed, all).
        label: Filter by label.
        assignee: Filter by assignee.
        limit: Maximum number of issues to return.
    """

    state: Optional[Literal["open", "closed", "all"]] = None
    label: Optional[str] = None
    assignee: Optional[str] = None
    limit: Optional[int] = Field(default=None, gt=0)


class Issue(BaseModel):
    """Represents a GitHub issue.

    Attributes:
        number: Issue number.
        title: Issue title.
        state: Issue state (open or closed).
        url: URL to the issue.
    """

    number: int
    title: str
    state: Literal["open", "closed"]
    url: str


class PrCreateOptions(BaseModel):
    """Options for creating a GitHub pull request.

    Attributes:
        title: PR title.
        body: PR body/description.
        head: Source branch.
        base: Target branch (defaults to default branch).
        draft: Create as draft PR.
    """

    title: str
    body: str
    head: str
    base: Optional[str] = None
    draft: bool = False


class PullRequest(BaseModel):
    """Represents a GitHub pull request.

    Attributes:
        number: PR number.
        title: PR title.
        state: PR state (open, closed, or merged).
        url: URL to the PR.
    """

    number: int
    title: str
    state: Literal["open", "closed", "merged"]
    url: str


class GitHubConnectorOptions(BaseModel):
    """Options for GitHubConnector.

    Attributes:
        debug: Enable debug logging of commands (never logs stdout for security).
    """

    debug: bool = False


# ═══════════════════════════════════════════════════════════════════════════════
# GITHUB CONNECTOR
# ═══════════════════════════════════════════════════════════════════════════════


class GitHubConnector:
    """Connector for interacting with GitHub via the gh CLI.

    SECURITY: This connector NEVER logs stdout to prevent accidental
    exposure of tokens or sensitive data from gh CLI output.

    Example:
        >>> gh = GitHubConnector(GitHubConnectorOptions(debug=True))
        >>>
        >>> # Create an issue
        >>> issue_num = await gh.issue_create(IssueCreateOptions(
        ...     title="Bug report",
        ...     body="Description of the bug",
        ...     labels=["bug"],
        ... ))
        >>>
        >>> # List issues
        >>> issues = await gh.issue_list("owner/repo", IssueFilters(state="open", limit=10))
    """

    def __init__(self, options: Optional[GitHubConnectorOptions] = None) -> None:
        self._options = options

    async def issue_create(self, opts: IssueCreateOptions) -> int:
        """Create a new GitHub issue.

        Args:
            opts: Issue creation options.

        Returns:
            Issue number on success.

        Raises:
            RuntimeError: If the gh command fails or output cannot be parsed.

        Example:
            >>> num = await gh.issue_create(IssueCreateOptions(
            ...     title="Bug", body="Details"
            ... ))
        """
        cmd: list[str] = ["gh", "issue", "create", "--title", opts.title, "--body", opts.body]

        if opts.repo:
            cmd.extend(["--repo", opts.repo])

        if opts.labels:
            for label in opts.labels:
                cmd.extend(["--label", label])

        if opts.project:
            cmd.extend(["--project", opts.project])

        result = await self._exec_gh(cmd)

        if is_exec_error(result):
            raise RuntimeError(f"Failed to create issue: {result.stderr or result.error_code}")

        # Parse issue number from URL (e.g., https://github.com/owner/repo/issues/123)
        import re

        match = re.search(r"/issues/(\d+)", result.stdout)
        if not match:
            raise RuntimeError("Could not parse issue number from gh output")

        return int(match.group(1))

    async def issue_list(
        self, repo: str, filters: Optional[IssueFilters] = None
    ) -> list[Issue]:
        """List issues from a repository.

        Args:
            repo: Repository in owner/repo format.
            filters: Optional filters for state, label, assignee, limit.

        Returns:
            List of Issue objects.

        Raises:
            RuntimeError: If the gh command fails.

        Example:
            >>> issues = await gh.issue_list("owner/repo", IssueFilters(state="open"))
        """
        cmd: list[str] = [
            "gh", "issue", "list",
            "--repo", repo,
            "--json", "number,title,state,url",
        ]

        if filters:
            if filters.state:
                cmd.extend(["--state", filters.state])
            if filters.label:
                cmd.extend(["--label", filters.label])
            if filters.assignee:
                cmd.extend(["--assignee", filters.assignee])
            if filters.limit:
                cmd.extend(["--limit", str(filters.limit)])

        result = await self._exec_gh(cmd)

        if is_exec_error(result):
            raise RuntimeError(f"Failed to list issues: {result.stderr or result.error_code}")

        raw_issues = json.loads(result.stdout)

        return [
            Issue(
                number=issue["number"],
                title=issue["title"],
                state=issue["state"].lower(),
                url=issue["url"],
            )
            for issue in raw_issues
        ]

    async def pr_create(self, opts: PrCreateOptions) -> int:
        """Create a new pull request.

        Args:
            opts: PR creation options.

        Returns:
            PR number on success.

        Raises:
            RuntimeError: If the gh command fails or output cannot be parsed.

        Example:
            >>> num = await gh.pr_create(PrCreateOptions(
            ...     title="Fix bug", body="Details", head="feature-branch"
            ... ))
        """
        cmd: list[str] = [
            "gh", "pr", "create",
            "--title", opts.title,
            "--body", opts.body,
            "--head", opts.head,
        ]

        if opts.base:
            cmd.extend(["--base", opts.base])

        if opts.draft:
            cmd.append("--draft")

        result = await self._exec_gh(cmd)

        if is_exec_error(result):
            raise RuntimeError(f"Failed to create PR: {result.stderr or result.error_code}")

        # Parse PR number from URL (e.g., https://github.com/owner/repo/pull/123)
        import re

        match = re.search(r"/pull/(\d+)", result.stdout)
        if not match:
            raise RuntimeError("Could not parse PR number from gh output")

        return int(match.group(1))

    async def pr_list(
        self, repo: str, filters: Optional[IssueFilters] = None
    ) -> list[PullRequest]:
        """List pull requests from a repository.

        Args:
            repo: Repository in owner/repo format.
            filters: Optional filters (reuses IssueFilters for similar fields).

        Returns:
            List of PullRequest objects.

        Raises:
            RuntimeError: If the gh command fails.

        Example:
            >>> prs = await gh.pr_list("owner/repo", IssueFilters(state="open"))
        """
        cmd: list[str] = [
            "gh", "pr", "list",
            "--repo", repo,
            "--json", "number,title,state,url",
        ]

        if filters:
            if filters.state:
                cmd.extend(["--state", filters.state])
            if filters.label:
                cmd.extend(["--label", filters.label])
            if filters.assignee:
                cmd.extend(["--assignee", filters.assignee])
            if filters.limit:
                cmd.extend(["--limit", str(filters.limit)])

        result = await self._exec_gh(cmd)

        if is_exec_error(result):
            raise RuntimeError(f"Failed to list PRs: {result.stderr or result.error_code}")

        raw_prs = json.loads(result.stdout)

        return [
            PullRequest(
                number=pr["number"],
                title=pr["title"],
                state=pr["state"].lower(),
                url=pr["url"],
            )
            for pr in raw_prs
        ]

    async def _exec_gh(self, cmd: list[str]) -> "ExecResult":
        """Execute a gh command.

        SECURITY: Only passes debug flag for command logging, never logs stdout.
        """
        from afd.platform import ExecResult

        exec_opts = ExecOptions()
        if self._options and self._options.debug:
            exec_opts.debug = True
        return await exec_command(cmd, exec_opts)
